### PR TITLE
Fix #37: Disable non-solution keyboard characters in Jumble

### LINKED ISSUE #37: [FEATURE] Disable non-solution characters from Jumble
**Is your feature request related to a problem? Please describe.**

Jumble currently allows **all** characters to be typed using keyboard and on-screen keyboard. This is not ideal, as most characters will never be part of the solution.

**Describe the solution you'd like**

- The input for not to allow non-solution characters for entry. A beep/jiggle or something to signify invalid input (This is currently done only at the end of submission).
- The on-screen keyboard to visually show that the character cannot be typed (button disabled).

**Describe alternatives you've considered**

N/A

**Additional context**

| State of things | Example solution |
|----------------|------------------|
| Screenshot 2023-10-14 at 6 32 13 PM | Screenshot 2023-10-14 at 6 31 58 PM |

### PR DESCRIPTION
**Pull request by paulkim26**
Fixes #37 

- Non-solution characters in Jumble's visual keyboard will appear faded out/disabled.
  - Added a new `.customButton[disabled]` CSS class in [global.css](https://github.com/paulkim26/ikea-word-games/blob/5e4034ee01797861d7cb9c5b56d73549dbdc2c89/src/app/global.css#L44).
  - In [WordInput.tsx](https://hodgef.com/simple-keyboard/documentation/options/buttonattributes/) a `disabled` attribute is applied to non-solution keys i.e. keys not found in the [targetValue](https://github.com/paulkim26/ikea-word-games/blob/5e4034ee01797861d7cb9c5b56d73549dbdc2c89/src/components/wordInput.tsx#L35) property of this component.
  - Note: A limitation of the `simple-keyboard` [buttonAttributes](https://hodgef.com/simple-keyboard/documentation/options/buttonattributes/) option is that both an attribute AND a value must be assigned. This means that the disabled keys feature the non-standard attribute `disabled="true"` vs. the standard `disabled` (which is why I could not leverage the `.customButton:disabled` pseudoclass).
- Jumble input field will jiggle if the player clicks on a non-solution character.
  - [onKeyboardInputChange](https://github.com/paulkim26/ikea-word-games/blob/5e4034ee01797861d7cb9c5b56d73549dbdc2c89/src/components/wordInput.tsx#L93) will check whether "disabled" characters are found in the keyboard component's input. This callback function will return early if so.
  - My original approach was to disallow disabled characters via the [inputPattern](https://hodgef.com/simple-keyboard/documentation/options/inputpattern/) option, but this would not allow for a callback function (i.e. to jiggle the input field) to fire.

If you see any issues with my pull request or suggestions for improvement, please let me know.

Tags: `hacktoberfest`, `bug`

### CODE REVIEW

File: src/app/global.css [hunk 1]
```diff
@@ -41,6 +41,10 @@ body {
 	border-color: #C4C4C4;
 }
 
+.simple-keyboard.hg-layout-default .customButton[disabled] {
+	opacity: 0.5;
```
Comment (avikantz) [inline src/app/global.css:45]:
```suggestion
	opacity: 0.5;
	cursor: not-allowed;
```

**Review by avikantz — COMMENTED** [review]
Looks good, please see comments for some minor suggestions.

File: src/components/wordInput.tsx [hunk 2]
```diff
@@ -55,6 +55,11 @@ export const WordInput = forwardRef<HTMLInputElement, WordInputProps>((props, in
     onClose: hideKeyboard,
   } = useDisclosure();
 
+  const enabledLetters = targetValue.removeAccents().split("");
+  const disabledLetters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ".split("").filter((letter: string) => {
+    return !enabledLetters.includes(letter);
+  });
```
[the remark this answers is not in the document]
Comment (paulkim26) [reply src/components/wordInput.tsx:61]: Thanks for the suggestion, will add the new utility function & remove redundant instances of split and join.

File: src/components/wordInput.tsx [hunk 3]
```diff
@@ -55,6 +55,11 @@ export const WordInput = forwardRef<HTMLInputElement, WordInputProps>((props, in
     onClose: hideKeyboard,
   } = useDisclosure();
 
+  const enabledLetters = targetValue.removeAccents().split("");
```
[the remark this answers is not in the document]
Comment (paulkim26) [reply src/components/wordInput.tsx:58]: Ah good catch, thanks.

File: src/components/wordInput.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (paulkim26) [reply src/components/wordInput.tsx]: I've created a new utils function and removed all instances of split & join.

File: src/components/wordInput.tsx [hunk 4]
```diff
@@ -206,6 +220,13 @@ export const WordInput = forwardRef<HTMLInputElement, WordInputProps>((props, in
               class: "actionButton",
             },
           ]}
+          buttonAttributes={[
+            {
+              attribute: "disabled",
+              value: "true",
+              buttons: disabledLetters.replace(/./g, '$& '),
```
Comment (paulkim26) [inline src/components/wordInput.tsx:227]: simple-keyboard needed spaces between each letter in order to parse this string

Comment (avikantz) [reply src/components/wordInput.tsx:229]: Ah, got it; this works as well

**Review by avikantz — APPROVED** [review]
Looks good. Thanks @paulkim26 🎉

**Comment by paulkim26** [comment]
Great to hear, thanks for the review!

**Pull request merged**